kong/resty/dns/client.lua

```diff
@@ -838,39 +811,45 @@ end
 local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
   local item = queue[key]
+  local is_yieldable = in_yieldable_phase()
 
-  -- if nothing is in progress, we start a new async query
-  if not item then
-    local err
-    item, err = asyncQuery(qname, r_opts, try_list)
+  -- If not yieldable, we start a new async query and return nil
+  if not is_yieldable then
     if not item then
-      return item, err, try_list
+      asyncQuery(qname, r_opts, try_list)
     end
-  else
-    add_status_to_try_list(try_list, "in progress (sync)")
+    -- item.result is definitely nil in this situation
+    return nil, "API disabled in the unyieldable phase", try_list
   end
 
-  local supported_semaphore_wait_phases = {
-    rewrite = true,
-    access = true,
-    content = true,
-    timer = true,
-    ssl_cert = true,
-    ssl_session_fetch = true,
-  }
+  -- If nothing is in progress and in the yieldable phase,
+  -- we start a new sync query
+  if not item then
+    item = {
+      key = key,
+      semaphore = semaphore(),
+      qname = qname,
+      r_opts = cycle_aware_deep_copy(r_opts),
+      try_list = try_list,
+    }
+    queue[key] = item
 
-  local ngx_phase = get_phase()
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
 
-  if not supported_semaphore_wait_phases[ngx_phase] then
-    -- phase not supported by `semaphore:wait`
-    -- return existing query (item)
-    --
-    -- this will avoid:
-    -- "dns lookup pool exceeded retries" (second try and subsequent retries)
-    -- "API disabled in the context of init_worker_by_lua" (first try)
-    return item, nil, try_list
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
```

Comment:
We should avoid copy&pasting.
At least this should be extracted into a function.

Reply:
And a possible way to refactor it:
Introduce a type named future, and a future object is returned by the call.
Caller calls future:get() to wait for the value (in which it waits for the semaphore and value);
Callee calls future:set() to post the semaphore and return the value.

Reply:
done , abstract common function for item creating and releasing